src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/core/client/RequestCompression.h>
+
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+#include <aws/core/utils/StringUtils.h>
+#include "zlib.h"
+
+#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(__CYGWIN__)
+#include <fcntl.h>
+#include <io.h>
+#define SET_BINARY_MODE(file) setmode(fileno(file), O_BINARY)
+#else
+#define SET_BINARY_MODE(file)
+#endif
+// Defining zlib chunks to be 256k
+#define ZLIB_CHUNK 263144
+#endif
+
+Aws::Utils::Outcome<std::shared_ptr<Aws::IOStream>, bool>
+Aws::Client::RequestCompression::compress(const Aws::String &input) const {
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+  // Preparing output
+  std::shared_ptr<Aws::IOStream> output =
+      Aws::MakeShared<Aws::StringStream>("REQUEST_COMPRESSION");
+
+  // Prepare ZLIB to compress
+  int ret;
+  z_stream strm;
+  unsigned char in[ZLIB_CHUNK];
```

Comment:
Be careful when allocating buffers on stack. Stack overflow may happen :-).

Reply:
It also is ignoring CUSTOM_MEMORY_ALLOCATORs here, I'm going to work on the memory model in a follow up PR. This is a first very early iteration of what would be at least 4 PRs. Just want to keep it functional.

Reply:
No longer in the stack, using unique pointers to new memory.